hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java

```diff
@@ -33,323 +32,379 @@
+					objectArrays.pop();
+					break;
+				case NULL_VALUE:
+					if ( subArrayObjectList != null ) {
+						// dealing with arrays
+						subArrayObjectList.add( null );
+					}
+					else {
+						objectArrays.getCurrent()[currentSelectableIndexInResultArray] = null;
+					}
+					break;
+				case NUMERIC_VALUE:
+					if ( subArrayObjectList != null ) {
+						// dealing with arrays
+						subArrayObjectList.add( adapter.fromNumericValue( subArrayObjectTypes.getElementType().getJdbcJavaType(),
+								subArrayObjectTypes.getElementType().getJdbcType(),reader,options));
+					}
+					else {
+						objectArrays.getCurrent()[currentSelectableIndexInResultArray] = adapter.fromNumericValue( currentSelectableMapping.getJdbcMapping().getJdbcJavaType(),
+								currentSelectableMapping.getJdbcMapping().getJdbcType(),reader,options);
+					}
+					break;
+				case BOOLEAN_VALUE:
+					if ( subArrayObjectList != null ) {
+						// dealing with arrays
+						subArrayObjectList.add( reader.getBooleanValue()?Boolean.TRUE:Boolean.FALSE);
+					}
+					else {
+						objectArrays.getCurrent()[currentSelectableIndexInResultArray] = reader.getBooleanValue()?Boolean.TRUE:Boolean.FALSE;
+					}
+					break;
+				case VALUE:
+					if ( subArrayObjectList != null ) {
+						// dealing with arrays
+						subArrayObjectList.add(adapter.fromValue( subArrayObjectTypes.getElementType().getJdbcJavaType(),
+								subArrayObjectTypes.getElementType().getJdbcType(),reader,options));
+					}
+					else {
+						objectArrays.getCurrent()[currentSelectableIndexInResultArray] = adapter.fromValue( currentSelectableMapping.getJdbcMapping().getJdbcJavaType(),
+								currentSelectableMapping.getJdbcMapping().getJdbcType(),reader,options);
+					}
+
+					break;
+				default:
+					assert false: "Unexpected type " + type;
+			}
 		}
+		return (X) objectArrayResult;
 	}
 
-	public static <X> X fromString(
+	/**
+	 * Deserialize a JSON value to Java Object
+	 * @param embeddableMappingType the mapping type
+	 * @param reader the JSON reader
+	 * @param returnEmbeddable do we return an Embeddable object or array of Objects
+	 * @param options wrappping options
+	 * @return the deserialized value
+	 * @param <X>
+	 * @throws SQLException
+	 */
+	public static <X> X deserialize(
 			EmbeddableMappingType embeddableMappingType,
-			String string,
+			JsonDocumentReader reader,
 			boolean returnEmbeddable,
 			WrapperOptions options) throws SQLException {
-		if ( string == null ) {
-			return null;
-		}
 
-		final int jdbcValueCount = embeddableMappingType.getJdbcValueCount();
-		final Object[] values = new Object[jdbcValueCount + ( embeddableMappingType.isPolymorphic() ? 1 : 0 )];
-		final int end = fromString( embeddableMappingType, string, 0, string.length(), values, returnEmbeddable, options );
-		assert string.substring( end ).isBlank();
+
+		final Object[] values = consumeJsonDocumentItems(reader, embeddableMappingType, returnEmbeddable, options);
```

Comment:
Let's add some proper checks. The first event read from the `reader` should be `OBJECT_START`.

```suggestion
		final JsonDocumentItemType event;
		if ( !reader.hasNext() || ( event = reader.next() ) == NULL_VALUE ) {
			return null;
		}
		if ( event != OBJECT_START ) {
			throw new IllegalArgumentException("Malformed JSON. Expected object but got: " + event);
		}
		final Object[] values = consumeJsonDocumentItems(reader, embeddableMappingType, returnEmbeddable, options);
```